File: implementation/src/main/java/ro/anud/xml_xsd/implementation/model/WorldStep/RuleGroup/LocationGraphRule/NodeRule/LinkGroupList/LinkGroup/LinkGroup.java
package ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.LocationGraphRule.NodeRule.LinkGroupList.LinkGroup;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import ro.anud.xml_xsd.implementation.util.RawNode;

import java.util.*;
import ro.anud.xml_xsd.implementation.util.Subscription;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static ro.anud.xml_xsd.implementation.util.LocalLogger.logEnter;
import static ro.anud.xml_xsd.implementation.util.LocalLogger.logReturn;
import static ro.anud.xml_xsd.implementation.util.LocalLogger.logReturnVoid;

  @EqualsAndHashCode
  @ToString
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
  public class LinkGroup implements  ro.anud.xml_xsd.implementation.model.interfaces.IType_linkGroup.IType_linkGroup<LinkGroup>,  ro.anud.xml_xsd.implementation.util.LinkedNode {

    public static String nodeName = "link_group";
    public static LinkGroup fromRawNode(RawNode rawNode) {
      logEnter();
      var instance = new LinkGroup();
      instance.rawNode(rawNode);
      instance.deserialize(rawNode);
      return logReturn(instance);
    }
    public static LinkGroup fromRawNode(RawNode rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      logEnter();
      var instance = fromRawNode(rawNode);
      instance.parentNode(parent);
      return logReturn(instance);
    }
    public static Optional<LinkGroup> fromRawNode(Optional<RawNode> rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
        logEnter();
        return logReturn(rawNode.map(o -> LinkGroup.fromRawNode(o, parent)));
    }
    public static List<LinkGroup> fromRawNode(List<RawNode> rawNodeList, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      logEnter();
      List<LinkGroup> returnList = Optional.ofNullable(rawNodeList)
          .orElse(List.of())
          .stream()
          .map(o -> LinkGroup.fromRawNode(o, parent))
          .collect(Collectors.toList());
      return logReturn(returnList);
    }

    public String classTypeId() {
      return ".world_step.rule_group.location_graph_rule.node_rule.link_group_list.link_group";
    }

    //Attributes

    //Attributes of type__link_group

    private String id;

    private Integer angle;
    @Builder.Default
    private Optional<Integer> angleMax = Optional.empty();
    @Builder.Default
    private Optional<Integer> limit = Optional.empty();

    //Children elements

    //Children of type__link_group
    @Builder.Default
    private List<ro.anud.xml_xsd.implementation.model.Type_linkGroup.ToOption.ToOption> toOption = new ArrayList<>();

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private RawNode rawNode = new RawNode();

    public RawNode rawNode() {
      return rawNode;
    }
    public void rawNode(RawNode rawNode) {
      this.rawNode = rawNode;
    }

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode = Optional.empty();

    public Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode() {
      return parentNode;
    }

    @Builder.Default
    private List<Consumer<List<Object>>> onChangeList = new ArrayList<>();

    public String nodeName() {
      return "link_group";
    }

    public void notifyChange(List<Object> list) {
      list.addLast(this);
      onChangeList.forEach(consumer -> consumer.accept(list));
      parentNode.ifPresent(linkedNode -> linkedNode.notifyChange(list));
    }

    public void parentNode(ro.anud.xml_xsd.implementation.util.LinkedNode linkedNode) {
      this.parentNode = Optional.of(linkedNode);
      notifyChange();
    }

    public Optional<ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.LocationGraphRule.NodeRule.LinkGroupList.LinkGroupList> parentAsLinkGroupList() {
      return parentNode.flatMap(node -> {
       if (node instanceof ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.LocationGraphRule.NodeRule.LinkGroupList.LinkGroupList casted){
         return Optional.of(casted);
       }
       return Optional.empty();
     });
    }

    public void removeChild(Object object) {
    }

    public int buildIndexForChild(Object object) {
        return 0;
    }

    public void removeFromParent() {
      parentNode.ifPresent(node -> node.removeChild(this));
    }

    public Subscription onChange(Consumer<List<Object>> onChange) {
      logEnter();
      onChangeList.add(onChange);
      return logReturn(() -> onChangeList.remove(onChange));
    }

    public void deserialize (RawNode rawNode) {
      var logger = logEnter();
      this.rawNode = rawNode;
      // Godot.GD.Print("Deserializing link_group");

      var innerLogger = logger.log("attributes");
      //Deserialize attributes

      // Deserialize arguments of type__link_group
      innerLogger.log("id");
      this.id = rawNode.getAttributeRequired("id");
      innerLogger.log("angle");
      this.angle = rawNode.getAttributeIntRequired("angle");
      innerLogger.log("angleMax");
      this.angleMax = rawNode.getAttributeInt("angleMax");
      innerLogger.log("limit");
      this.limit = rawNode.getAttributeInt("limit");
      innerLogger = logger.log("children");
      //Deserialize children

      // Deserialize children of type__link_group
      this.toOption = ro.anud.xml_xsd.implementation.model.Type_linkGroup.ToOption.ToOption.fromRawNode(rawNode.getChildrenList("to_option"), this);
      logReturnVoid();
    }

    public RawNode serializeIntoRawNode()
    {
      var logger = logEnter();
      rawNode.setTag("link_group");
      var innerLogger = logger.log("attributes");
      //Serialize attributes

      // Serialize arguments of type__link_group
      innerLogger.log("id");
      rawNode.setAttribute("id", this.id);
      innerLogger.log("angle");
      rawNode.setAttribute("angle", this.angle);
      innerLogger.log("angleMax");
      this.angleMax.ifPresent(o -> rawNode.setAttribute("angleMax", o));
      innerLogger.log("limit");
      this.limit.ifPresent(o -> rawNode.setAttribute("limit", o));

      innerLogger = logger.log("children");
      //Serialize children

      // Serialize children of type__link_group
      innerLogger.log("to_option");
      rawNode.setChildren("to_option", toOption.stream().map(ro.anud.xml_xsd.implementation.model.Type_linkGroup.ToOption.ToOption::serializeIntoRawNode).toList());
      return rawNode;
    }

    public void serialize(Document document, Element element)
    {
        // Godot.GD.Print("Serializing link_group");
        var updatedRawNode = serializeIntoRawNode();
        updatedRawNode.populateNode(document, element);
    }
    public String getId()
    {
      return this.id;
    }
    public LinkGroup setId(String value)
    {
      this.id = value;
      notifyChange();
      return this;
    }
    public Integer getAngle()
    {
      return this.angle;
    }
    public LinkGroup setAngle(Integer value)
    {
      this.angle = value;
      notifyChange();
      return this;
    }
    public Optional<Integer> getAngleMax()
    {
      return this.angleMax;
    }
    public LinkGroup setAngleMax(Optional<Integer> value)
    {
      this.angleMax = value;
      notifyChange();
      return this;
    }
    public Optional<Integer> getLimit()
    {
      return this.limit;
    }
    public LinkGroup setLimit(Optional<Integer> value)
    {
      this.limit = value;
      notifyChange();
      return this;
    }
    public List<ro.anud.xml_xsd.implementation.model.Type_linkGroup.ToOption.ToOption> getToOption()
    {
      return this.toOption;
    }
    public java.util.stream.Stream<ro.anud.xml_xsd.implementation.model.Type_linkGroup.ToOption.ToOption> streamToOption()
    {
      return toOption.stream();
    }
    public LinkGroup addToOption(ro.anud.xml_xsd.implementation.model.Type_linkGroup.ToOption.ToOption value)
    {
      this.toOption.add(value);
      value.parentNode(this);
      notifyChange();
      return this;
    }
    public LinkGroup addAllToOption(List<ro.anud.xml_xsd.implementation.model.Type_linkGroup.ToOption.ToOption> value)
    {
      this.toOption.addAll(value);
      value.forEach(e -> e.parentNode(this));
      notifyChange();
      return this;
    }
    public LinkGroup removeToOption(ro.anud.xml_xsd.implementation.model.Type_linkGroup.ToOption.ToOption value)
    {
      this.toOption.remove(value);
      notifyChange();
      return this;
    }

    public ro.anud.xml_xsd.implementation.util.LinkedNode deserializeAtPath(String xpath, RawNode rawNode) {
       if(xpath.startsWith("."))
        {
          xpath = xpath.substring(1);
        }

        deserialize(rawNode);
        return this;
    }
  }


  /*
    dependant type:
    {
      "type": "element",
      "value": {
        "metaType": "object",
        "attributes": {
          "metaType": "object",
          "value": {}
        },
        "value": {}
      },
      "name": "link_group"
    }
  */

File: gui-client/world_step/world_step/rule_group/location_graph_rule/node_rule/link_group_list/link_group/link_group.cs
using System;
using System.Collections.Immutable;
using System.Collections.Generic;
using System.Xml;
using System.Linq;
using Guiclient.util;
using Godot;
using XSD;

namespace XSD.Nworld_step.Nrule_group.Nlocation_graph_rule.Nnode_rule.Nlink_group_list.Nlink_group {}
namespace XSD {
}
namespace XSD.Nworld_step.Nrule_group.Nlocation_graph_rule.Nnode_rule.Nlink_group_list {
  public class link_group : IEquatable<link_group>, XSD.ILinkedNode , Itype__link_group {

    public static string ClassTypeId = ".world_step.rule_group.location_graph_rule.node_rule.link_group_list.link_group";
    public static string TagName = "link_group";

    public string NodeName {get =>"link_group";}
    public RawNode rawNode = new RawNode();

    private ILinkedNode? _parentNode;
    public ILinkedNode? ParentNode {get => _parentNode; set => _parentNode = value;}
    private List<Action<link_group>> _onSelfChangeCallbackList = new();
    private List<Action<List<ILinkedNode>>> _onChangeCallbackList = new();

    //Attributes

    //Attributes of type__link_group
    private System.String _id;
    public System.String id { get => _id; set => _id = value; }
    private System.Int32 _angle;
    public System.Int32 angle { get => _angle; set => _angle = value; }
    private System.Int32? _angleMax;
    public System.Int32? angleMax { get => _angleMax; set => _angleMax = value; }
    private System.Int32? _limit;
    public System.Int32? limit { get => _limit; set => _limit = value; }

    //Children elements

    //Children of type__link_group

    private LinkedNodeCollection<XSD.Ntype__link_group.to_option> _to_option = new();
    public LinkedNodeCollection<XSD.Ntype__link_group.to_option> to_option
    {
      get => _to_option;
      set
      {
        _to_option = value;
        value.ForEach(linkedNode => linkedNode.ParentNode = this);
        _to_option.OnAdd = (value) =>
        {
          value.ParentNode = this;
          NotifyChange();
        };
      }
    }
    public link_group()
    {
    }

    public link_group(RawNode rawNode)
    {
      Deserialize(rawNode);
    }

    public link_group(XmlElement xmlElement)
    {
      this.rawNode.Deserialize(xmlElement);
      Deserialize(rawNode);
    }

    public void SetAttribute(string name, string? value)
    {
    }

    public void SetChild(dynamic linkedNode)
    {
    }

    public void ClearChild(dynamic linkedNode)
    {
    }

    public Action OnSelfChange(Action<link_group> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }

    public Action OnSelfChangeNode(Action<ILinkedNode> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }


    public Action OnChange(Action<List<ILinkedNode>> callback)
    {
      _onChangeCallbackList.Add(callback);
      return () => _onChangeCallbackList.Remove(callback);
    }

    public void Deserialize (RawNode rawNode)
    {
      this.rawNode = rawNode;
      // Godot.GD.Print("Deserializing link_group");
      //Deserialize arguments

      // Deserialize arguments of type__link_group
  if(rawNode.attributes.ContainsKey("id"))
  {
    var attribute_id = rawNode.attributes["id"];
    this.id = rawNode.attributes["id"];
  }
  if(rawNode.attributes.ContainsKey("angle"))
  {
    var attribute_angle = rawNode.attributes["angle"];
    this.angle = attribute_angle.ToInt();
  }
  if(rawNode.attributes.ContainsKey("angleMax"))
  {
    var attribute_angleMax = rawNode.attributes["angleMax"];
    this.angleMax = attribute_angleMax?.ToInt();
  }
  if(rawNode.attributes.ContainsKey("limit"))
  {
    var attribute_limit = rawNode.attributes["limit"];
    this.limit = attribute_limit?.ToInt();
  }

      //Deserialize children

      // Deserialize children of type__link_group
  to_option = rawNode.InitializeWithRawNode("to_option", to_option);
  to_option.OnAdd = (value) =>
    {
      value.ParentNode = this;
      NotifyChange();
    };
      NotifyChange();
    }

    public RawNode SerializeIntoRawNode()
    {
      //Serialize arguments

      // Serialize arguments of type__link_group
  if(this._id != null)
  {
    rawNode.attributes["id"] = this._id.ToString();
  }
  if(this._angle != null)
  {
    rawNode.attributes["angle"] = this._angle.ToString();
  }
  if(this._angleMax != null)
  {
    rawNode.attributes["angleMax"] = this._angleMax?.ToString();
  }
  if(this._limit != null)
  {
    rawNode.attributes["limit"] = this._limit?.ToString();
  }

      //Serialize children

      // Serialize children of type__link_group
  rawNode.children["to_option"] = to_option.Select(x => x.SerializeIntoRawNode()).ToList();
      return rawNode;
    }

    public void Serialize(XmlElement element)
    {
        // Godot.GD.Print("Serializing link_group");
        var updatedRawNode = SerializeIntoRawNode();
        updatedRawNode.Serialize(element);
    }
    public System.String Get_id()
    {
      return this.id;
    }
    public void Set_id(System.String value)
    {
      this.id = value;
      this.NotifyChange();
    }
    public System.Int32 Get_angle()
    {
      return this.angle;
    }
    public void Set_angle(System.Int32 value)
    {
      this.angle = value;
      this.NotifyChange();
    }
    public System.Int32? Get_angleMax()
    {
      return this.angleMax;
    }
    public void Set_angleMax(System.Int32? value)
    {
      this.angleMax = value;
      this.NotifyChange();
    }
    public System.Int32? Get_limit()
    {
      return this.limit;
    }
    public void Set_limit(System.Int32? value)
    {
      this.limit = value;
      this.NotifyChange();
    }


    public void DeserializeAtPath(string xpath, RawNode rawNode)
    {
      if(xpath.StartsWith("."))
      {
        xpath = xpath.Substring(1);
      }

      Deserialize(rawNode);
    }

    public void NotifyChange(List<ILinkedNode> linkedNodes)
    {
      if(_parentNode == null)
        return;
      linkedNodes.Add(this);
      _onSelfChangeCallbackList.ForEach(action => action(this));
      _onChangeCallbackList.ForEach(action => action(linkedNodes));
      _parentNode.NotifyChange(linkedNodes);
    }

    public void NotifyChange()
    {
      NotifyChange(new ());
    }

    public int? BuildIndexForChild(ILinkedNode linkedNode)
    {
      return null;
    }

    public bool IsValidChildType(ILinkedNode candidateChild) {
      return false;
    }

        public bool Equals(link_group? obj)
        {
            if (obj == null || GetType() != obj.GetType())
                return false;

            var other = (link_group)obj;
            return object.Equals(this, other);
        }

    public int GetHashCode()
    {
        return base.GetHashCode();
    }
  }
}